packages/cli/src/s3-store-wrapper.ts

```diff
@@ -0,0 +1,101 @@
+import { StoreWrapperInterface, StoreSearchParams } from '@ceramicnetwork/core'
+import LevelUp from 'levelup'
+import S3LevelDOWN from 's3leveldown'
+import toArray from 'stream-to-array'
+import PQueue from 'p-queue'
+
+/**
+ * Maximum GET/HEAD requests per second to AWS S3
+ */
+const MAX_LOAD_RPS = 4000
+
+export class S3StoreWrapper implements StoreWrapperInterface {
+  readonly networkName: string
+  readonly #bucketName: string
+  #store:  LevelUp.LevelUp
+  #subStores: Record<string, LevelUp.LevelUp> = {}
+
+  readonly  #loadingLimit = new PQueue({
+    intervalCap: MAX_LOAD_RPS,
+    interval: 1000,
+    carryoverConcurrencyCount: true,
+  })
+
+  constructor(bucketName: string, networkName: string) {
+    this.#bucketName = bucketName
+    this.networkName = networkName
+    const location = this.#bucketName + '/ceramic/' + this.networkName + '/state-store'
+    // @ts-ignore
+    this.#store = new LevelUp(new S3LevelDOWN(location))
+  }
+
+  private throwIfNotInitialized(): void {
+    if (!this.#store) throw new Error('You must call async init(), before you start using the S3Store')
+  }
+
+  private getLocation(subChannel?: string) {
+    const baseLocation = this.#bucketName + '/ceramic/' + this.networkName + '/state-store'
+    return subChannel ? `${baseLocation}/${subChannel}` : baseLocation
+  }
+
+  private getStore(subChannel?: string) {
```

Comment:
IMO, optional `subChannel` location could be handled better, if you do not have a division between `this.#store` and `this.#subStores`. You could enumerate the stores by full path, instead of `subChannel` name. In this case `this.#store` becomes `this.#subStores(this.getLocation(undefined))`.

And, you could simplify code even further by defining a `SublevelsMap` class and using it instead of current `this.#subStores`:
```
class SublevelsMap {
  #map: Map<string, LevelUp.LevelUp>
  #baseLocation: string
  constructor(baseLocation: string) {
    this.#baseLocation = baseLocation
    this.#map.set(this.#baseLocation, new LevelUp(...))
  }
  
  get(sublevelName?: string) {
    return this.#map.get(this.fullLocation(sublevelName))
  }
  
  private fullLocation(sublevelName?: string) {
    if (sublevelName) {
      return `${this.#baseLocation}/${sublevelName}`
    } else {
      return this.#baseLocation
    }
  }
}
```

Reply:
Also, is this thing called a sublevel or a subchannel in LevelDB parlance?  I use sublevel here, based on our previous discussion, but I really do not know, if subchannel is a more appropriate name for this.

Reply:
@ukstv , do you think it makes sense to have two slightly different `Map` classes for S3 store and for LevelDB store?

Reply:
> @ukstv , do you think it makes sense to have two slightly different Map classes for S3 store and for LevelDB store?

I see from your comment below that you think it does make sense. Cool, I'm gonna implement these two map classes then :)